rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+export type Unmount = () => void;
+
+export interface AppSpec {
+  id: string;
+
+  /**
+   * The title of the application.
+   */
+  title: string;
+
+  /**
+   * The root route to mount this application at. Must be unique across all
+   * registered applications.
+   */
+  rootRoute: string;
+
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
+   * @param targetDomElement An HTMLElement to mount the application onto.
+   * @returns An unmounting function that will be called to unmount the application.
+   */
+  mount(targetDomElement: HTMLElement): Unmount
+
+  /**
+   * A EUI iconType that will be used for the app's icon. This icon
+   * takes precendence over the `icon` property.
+   */
+  euiIconType?: string;
+
+  /**
+   * A URL to an image file used as an icon. Used as a fallback
+   * if `euiIconType` is not provided.
+   */
+  icon?: string;
+
+  /**
+   * Custom capabilities defined by the app.
+   */
+  capabilities?: Partial<Capabilities>;
+}
+
+export interface ApplicationSetup {
+  /**
+   * Registers an application with the system.
+   */
+  registerApp(app: AppSpec): void;
+}
+
+export interface ApplicationStart {
+  /**
+   * The UI capabilities for the current user.
+   */
+  capabilities: Capabilties;
+}
+```
+
+## Mounting
+
+When an app is registered via `registerApp`, it must provide a `mount` function
+that will be invoked whenever the window's location has changed from another app
+to this app.
+
+This function is called with an `HTMLELement` for the application
+to render itself to. The application must also return a function that can be
+called by the ApplicationService to unmount the application at the given DOM
+node.
+
+The ApplicationService's `registerApp` method will only be available during the
+*setup* lifecycle event. This allows the system to know when all applications
+have been registered.
+
+However, the `mount` function will also get access to the `PluginStartContext`
+that is normally provided during the *start* lifecycle event, as well as the
+*start* contracts for any plugin dependencies. (Note: applications can only be
+mounted during *start*).
+
+## Routing
+
+The ApplicationService will serve as the global frontend router for Kibana,
+enabling Kibana to be a 100% single page application. However, the router will
+only manage top-level routes. Applications themselves will need to implement
+their own routing as subroutes of the top-level route.
+
+An example:
+- "MyApp" is registered with `rootRoute: '/myApp'`
+- User navigates from mykibana.com/app/home to mykibana.com/app/myApp
```

Comment:
Do we need to prefix the rootRoute with /app? Can we require it be specified in rootRoute if so?

Reply:
One benefit of having /app in the rootRoute is apps that render under different routes, like `/login`. Maybe this RFC is implying that those won’t be apps anymore, but I think that needs to be covered explicitly if so.

Reply:
One benefit I considered of keeping the `/app` prefix is that it gives us a bit of wiggle room for a few things:
- We know when the backend gets a `/app` request that we need to render the core bundle rather than a JSON API endpoint. This simplifies the routing logic on the backend.
- If we don't do this now and later decide we do, we risk breaking a lot more URLs and potentially in ways we can't anticipate because we don't know how 3rd party plugins are using this so we can't necessarily redirect them.